src/relay/connection.ts

```diff
@@ -276,33 +281,48 @@ class RelayConnection extends EventEmitter implements MultiaddrConnection {
     while (true) {
       let promises: Promise<SinkType>[] = []
 
+      let resolvedPromiseName
+
+      const pushPromise = (promise: Promise<SinkType>, name: string) => {
```

Comment:
this makes things indeed way simpler

Reply:
You're right, it can be simplified and refactored, but that was not the goal for this PR. First I want to establish a consistent behavior and cover it with tests and then try refactoring stuff without risk of breaking it.